vault/data_source_azure_access_credentials.go

```diff
@@ -0,0 +1,243 @@
+	}
+	log.Printf("[DEBUG] Read %q from Vault", path)
+
+	if secret == nil {
+		return fmt.Errorf("no role found at path %q", path)
+	}
+
+	clientID := secret.Data["client_id"].(string)
+	clientSecret := secret.Data["client_secret"].(string)
+
+	d.SetId(secret.LeaseID)
+	_ = d.Set("client_id", secret.Data["client_id"])
+	_ = d.Set("client_secret", secret.Data["client_secret"])
+	_ = d.Set("lease_id", secret.LeaseID)
+	_ = d.Set("lease_duration", secret.LeaseDuration)
+	_ = d.Set("lease_start_time", time.Now().Format(time.RFC3339))
+	_ = d.Set("lease_renewable", secret.Renewable)
+
+	subscriptionID := d.Get("subscription_id").(string)
+	tenantID := d.Get("tenant_id").(string)
+
+	// If we're not supposed to validate creds, or we don't have enough
+	// information to do it, there's nothing further to do here.
+	validateCreds := d.Get("validate_creds").(bool)
+	if !validateCreds {
+		// We're done.
+		return nil
+	}
+	if subscriptionID == "" || tenantID == "" {
+		return fmt.Errorf(`if 'validate_creds' is true, both 'subscription_id' and 'tenant_id' must also be provided`)
+	}
+
+	// Let's, test the credentials before returning them.
+	vnetClient := network.NewVirtualNetworksClient(subscriptionID)
+
+	// You can't directly pass in a client ID and secret.
+	// The closest thing you can do is provide them through the environment.
+	revert, err := setAndPrepareReversion(auth.TenantID, tenantID)
+	if err != nil {
+		return err
+	}
+	defer revert()
+
+	revert, err = setAndPrepareReversion(auth.ClientID, clientID)
+	if err != nil {
+		return err
+	}
+	defer revert()
+
+	revert, err = setAndPrepareReversion(auth.ClientSecret, clientSecret)
+	if err != nil {
+		return err
+	}
+	defer revert()
+
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return err
+	}
+	vnetClient.Authorizer = authorizer
+
+	// By default, this loop will test credentials every 7 seconds. It will
+	// require 8 sequential successes before returning them. It may retry for
+	// up to 20 minutes.
+	// This is because a colleague with Azure experience has seen an acceptable
+	// success rate with these settings. That colleague has also observed creds
+	// taking up to 15 minutes to fully propagate.
+	// However, this will vary widely depending on the size of a user's Azure
+	// Active Directory. Thus, these settings are fully configurable.
+	credValidationTimeoutSecs := d.Get("max_cred_validation_seconds").(int)
+	sequentialSuccessesRequired := d.Get("num_sequential_successes").(int)
+	secBetweenTests := d.Get("num_seconds_between_tests").(int)
+
+	startTime := time.Now()
+	endTime := startTime.Add(time.Duration(credValidationTimeoutSecs) * time.Second)
+
+	// This URL merely lists providers, and should return a 200 if the credentials
+	// have propagated to the Azure server receiving the call.
+	u := fmt.Sprintf("https://management.azure.com/subscriptions/%s/providers?api-version=2019-10-01", subscriptionID)
```

Comment:
We probably need to account for the different environment since they have different endpoints. (https://github.com/Azure/go-autorest/blob/master/autorest/azure/environments.go#L34).  The environment should already be configured elsewhere for Azure Secrets to work, so presumably we can grab the endpoint from existing config.

Reply:
@kalafut do you think I should also read the tenant ID and subscription ID from there? https://www.vaultproject.io/api/secret/azure/index.html#sample-response. I was kind of torn on it because on the one hand, it's easy and prevents drift. On the other hand, it's slightly magical?

Reply:
I didn't think of that before, but I think that makes sense. If you're generating creds for (sub,tenant,env), then I can't imagine why you'd want to, or be successful with, verifying them against something other than (sub,tenant,env). I suppose we can start with the magic approach and if we get a request for explicit config we could easily add it without breaking compatibility (it would just be an override). But I really can't think of the use, so I'd say pull them all from there.

Reply:
This was done in https://github.com/terraform-providers/terraform-provider-vault/pull/713/commits/d9389dc4c6b1175a483a6becd3e89e2e913a4d6f. I also found a more direct way to _only_ do the token refresh call, which I think is preferable over my previous approach because it doesn't require any permissions for some separate call.